x/smartaccount/ante/auth.go

```diff
@@ -0,0 +1,285 @@
+	}
+	account := signers[0].String()
+
+	setting, err := sad.sak.GetSetting(ctx, account)
+	if sdkerrors.ErrKeyNotFound.Is(err) {
+		// run through the default handlers for signature verification
+		newCtx, err := sad.defaultVerifySigDecorator(ctx, tx, simulate)
+		if err != nil {
+			return newCtx, err
+		}
+		// continue to the next handler after default signature verification
+		return next(newCtx, tx, simulate)
+	} else if err != nil {
+		return ctx, err
+	}
+
+	// Sender here is the account that signed the transaction
+	// Could be different from the account above (confusingly named signer)
+	signatures, err := sigTx.GetSignaturesV2()
+	if err != nil {
+		return ctx, err
+	}
+	if len(signatures) == 0 {
+		return ctx, sdkerrors.ErrNoSignatures.Wrap("no signatures found")
+	} else if len(signatures) > 1 {
+		// TODO: remove when support multi sig auth
+		return ctx, sdkerrors.ErrUnauthorized.Wrap("multiple signatures not supported")
+	}
+
+	signature := signatures[0]
+	signaturesBs := [][]byte{}
+
+	senderAddr, err := sdk.AccAddressFromHexUnsafe(signature.PubKey.Address().String())
+	if err != nil {
+		return ctx, err
+	}
+
+	acc, err := authante.GetSignerAcc(ctx, sad.ak, senderAddr)
+	if err != nil {
+		return ctx, err
+	}
+	var accNum uint64
+	if ctx.BlockHeight() != 0 {
+		accNum = acc.GetAccountNumber()
+	}
+
+	signerData := authsigning.SignerData{
+		Address:       senderAddr.String(),
+		ChainID:       ctx.ChainID(),
+		AccountNumber: accNum,
+		Sequence:      acc.GetSequence(),
+		PubKey:        signature.PubKey,
+	}
+
+	signatureBz, err := signatureDataToBz(signature.Data)
+	if err != nil {
+		return ctx, err
+	}
+	signedBytes, err := GetSignBytesArr(signature.PubKey, signerData, signature.Data, sad.signModeHandler, tx)
+	if err != nil {
+		return ctx, err
+	}
+	signaturesBs = append(signaturesBs, signatureBz...)
+
+	ctx = ctx.WithValue(types.ModuleName, setting)
+	if setting.Authorization != nil && len(setting.Authorization) > 0 {
+		success := false
+		for _, auth := range setting.Authorization {
+			authMsg := types.Authorization{
+				Sender:      senderAddr.String(),
+				Account:     account,
+				Data:        []byte(auth.InitMsg),
+				Signatures:  signaturesBs,
+				SignedBytes: signedBytes,
+			}
+			authMsgBz, err := json.Marshal(authMsg)
+			if err != nil {
+				return ctx, err
+			}
+			contractAddr, err := sdk.AccAddressFromBech32(auth.ContractAddress)
+			if err != nil {
+				return ctx, err
+			}
+			_, err = sad.wk.Sudo(ctx, contractAddr, authMsgBz)
+			// so long as one of the authorization is successful, we're good
+			if err == nil {
+				success = true
+				break
+			}
+			if err != nil && setting.Fallback {
+				return next(ctx, tx, simulate)
+			} else if err != nil {
+				return ctx, err
+			}
+		}
+		if !success {
+			return ctx, sdkerrors.ErrUnauthorized.Wrap("authorization failed")
+		}
+	} else if setting.Fallback {
```

Comment:
Logic here should be
1. If there are no custom auths: Always use default sig verify 
2. If there are custom auth: Success if any of the auth passes
3. If all custom auth fails: If fallback = true, use default sig verify
4. Else fail